Upgrades README improvements (#417)
ava-labs/teleporter: contracts/src/Teleporter/upgrades/README.md

- Version zero is an invalid version, and is used to indicate that a `TeleporterMessenger` contract has not been registered yet.
- Once a version number is registered in the registry, it cannot be changed, but a previous registered protocol address can be added to the registry with a new version. This is especially important in the case of a rollback to a previous Teleporter version, in which case the previous Teleporter contract address would need to be registered with a new version to the registry.

> ## Integrating  `TeleporterUpgradeable` into a dApp

Reviewer: ```suggestion
## Integrating `TeleporterUpgradeable` into a dApp
```

        ITeleporterMessenger teleporterMessenger = _getTeleporterMessenger();
```

#### Receiving from specific Teleporter versions
### Receiving from specific Teleporter versions

`TeleporterUpgradeable` also provides an initial implementation of [ITeleporterReceiver.receiveTeleporterMessage](../ITeleporterReceiver.sol) that ensures `_msgSender` is a `TeleporterMessenger` contract with a version greater than or equal to `minTeleporterVersion`. This supports the case where a dApp wants to upgrade to a new version of the `TeleporterMessenger` contract, but still wants to be able to receive messages from the old Teleporter contract.The dApp can override `_receiveTeleporterMessage` to implement its own logic for receiving messages from Teleporter contracts.

#### Pausing Teleporter version interactions
## Managing a TeleporterUpgradeable dApp

> Dapps that inherit from `TeleporterUpgradeable` can pause Teleporter interactions by calling `TeleporterUpgradeable.pauseTeleporterAddress`. This function prevents the contract from interacting with the paused Teleporter address when sending or receiving Teleporter messages.

Reviewer 2: nit: Perhaps add a sentence here to summarize what managing a TeleporterUpgradable dApp involves (updating versions, pausing interactions, etc.). Otherwise we would have a subsection heading directly following a section heading.

Author: Good call, added now

        ITeleporterMessenger teleporterMessenger = _getTeleporterMessenger();
```

### Receiving from specific Teleporter versions

`TeleporterUpgradeable` also provides an initial implementation of [ITeleporterReceiver.receiveTeleporterMessage](../ITeleporterReceiver.sol) that ensures `_msgSender` is a `TeleporterMessenger` contract with a version greater than or equal to `minTeleporterVersion`. This supports the case where a dApp wants to upgrade to a new version of the `TeleporterMessenger` contract, but still wants to be able to receive messages from the old Teleporter contract.The dApp can override `_receiveTeleporterMessage` to implement its own logic for receiving messages from Teleporter contracts.

## Managing a TeleporterUpgradeable dApp

dApps that implement `TeleporterUpgradeable` automatically use the latest Teleporter version registered with the `TeleporterRegistry`. Interaction with underlying `TeleporterMessenger` versions can be managed by setting the minimum Teleporter version, and pausing and unpausing specific versions.

### Managing the Minimum Teleporter version

> The `TeleporterUpgradeable` contract constructor saves the Teleporter registry in a state variable used by the inheriting dApp contract, and initializes a `minTeleporterVersion` to the highest `TeleporterMessenger` version registered in `TeleporterRegistry`. `minTeleporterVersion` is used to allow dApp's to specify the Teleporter versions allowed to interact with it.

Reviewer: ```suggestion
The `TeleporterUpgradeable` contract constructor saves the Teleporter registry in a state variable used by the inheriting dApp contract, and initializes a `minTeleporterVersion` to the highest `TeleporterMessenger` version registered in `TeleporterRegistry`. `minTeleporterVersion` is used to allow dApps to specify the Teleporter versions allowed to interact with it.
```

#### Updating `minTeleporterVersion`

> The `TeleporterUpgradeable.updateMinTeleporterVersion` function updates the `minTeleporterVersion` used to check which Teleporter versions can deliver messages to the dApp, and emits the `MinTeleporterVersionUpdated` event. The `updateMinTeleporterVersion` function should **ONLY** be called by the dApp when it completes delivery of messages from the old Teleporter contract, and now wants to update the `minTeleporterVersion` to only allow the new Teleporter version. By default, `updateMinTeleporterVersion` can only be called with a version greater than the current `minTeleporterVersion` and less than `latestVersion` in the Teleporter registry. So once this function is called, the dApp will no longer be able to receive messages from the old Teleporter contract version, unless the old version's Teleporter address was registered in the registry again with a new version.

Reviewer: ```suggestion
The `TeleporterUpgradeable.updateMinTeleporterVersion` function updates the `minTeleporterVersion` used to check which Teleporter versions can be used for sending and receiving messages. **Once the `minTeleporterVersion` is increased, any undelivered messages sent by other chains using older versions of Teleporter will never be able to be received**. The `updateMinTeleporterVersion` function can only be called with a version greater than the current `minTeleporterVersion` and less than `latestVersion` in the Teleporter registry.
```

A few nits here: people may want to explicitly use `updateMinTeleporterVersion` to prevent existing messages from ever being delivered (ie. if they exploit a vulnerability), and I think this makes it a bit more concise to be more clear.

> Example: Update the minimum Teleporter version to 2
>```

> >cast send <DAPP_ADDRESS> "updateMinTeleporterVersion(uint256)" 2 

Reviewer 3: should we include other required fields for the command like rpc-url, private key etc as examples

Reviewer 4: Technically those fields aren't strictly required. For example, the rpc url will be taken from the `ETH_RPC_URL` env var if not supplied in the command.

We may still want to include examples with them, though.

Author: I added a note to the top of this section that the examples may need to be modified to work with different wallet solutions

>```

### Pausing Teleporter version interactions

> Dapps that inherit from `TeleporterUpgradeable` can pause Teleporter interactions by calling `TeleporterUpgradeable.pauseTeleporterAddress`. This function prevents the dApp contract from interacting with the paused Teleporter address when sending or receiving Teleporter messages.

Reviewer 3: think we should clarify that Teleporter itself is not paused, and it's just the dApp's sending/receiving of messages

Author: I tried to achieve this by saying that "Teleporter interactions" are paused - can you suggest an alternate phrasing?

> Dapps that inherit from `TeleporterUpgradeable` can pause Teleporter interactions by calling `TeleporterUpgradeable.pauseTeleporterAddress`. This function prevents the dApp contract from interacting with the paused Teleporter address when sending or receiving Teleporter messages.

Reviewer: Capitalization nit: In some places we start sentences with "dApp" and others "Dapps"

> Example: Pause Teleporter version 3
>```
>versionThreeAddress=$(cast call <REGISTRY_ADDRESS> "getAddressFromVersion(uint256)(address)" 3)
>cast send <DAPP_ADDRESS> "pauseTeleporterAddress(address)" $versionThreeAddress
>```

#### Pause all Teleporter interactions

To pause all Teleporter interactions, `TeleporterUpgradeable.pauseTeleporterAddress` must be called for every Teleporter version from the `minTeleporterVersion` to the latest Teleporter version registered in `TeleporterRegistry`. The latest Teleporter version can be obtained by inspecting the public variable `TeleporterRegistry.latestVersion`. The `minTeleporterVersion` can be obtained by calling `TeleporterUpgradeable.getMinTeleporterVersion`.

> Example: Pause all registered Teleporter versions
>```
># Fetch the minimum Teleporter version
>minVersion=$(cast call <DAPP_ADDRESS> "getMinTeleporterVersion()(uint256)")
>
># Fetch the latest registered version
>latestVersion=$(cast call <REGISTRY_ADDRESS> "latestVersion()(uint256)")
>
># Pause all registered versions
>for ((version=minVersion; version<=latestVersion; version++))
>do

> >  versionAddress=$(cast call <REGISTRY_ADDRESS> "getAddressFromVersion(uint256)(address)" $version)

Reviewer: I think it's possible for there to be gaps in version numbers right? If so, we should continue here if the version address isn't registered.

Author: Modified the example script, and added a note in the above description.
